faunadb-common/src/main/java/com/faunadb/common/Connection.java

```diff
@@ -358,90 +352,82 @@ public void syncLastTxnTime(long newTxnTime) {
+    String urlEncodedKey = URLEncoder.encode(queryKey, StandardCharsets.UTF_8);
+    String urlEncodedValue = URLEncoder.encode(queryValue, StandardCharsets.UTF_8);
+    String query = urlEncodedKey + "=" + urlEncodedValue;
+    return new URI(oldUri.getScheme(), oldUri.getAuthority(), oldUri.getPath(),
+            oldUri.getQuery() == null ? query : oldUri.getQuery() + "&" + query, oldUri.getFragment());
   }
 
-  private CompletableFuture<FullHttpResponse> performRequest(final FullHttpRequest request, final Optional<Duration> requestQueryTimeout) {
+  private CompletableFuture<HttpResponse<String>> performRequest(String httpMethod, String path, Optional<JsonNode> body,
+                                                                 Map<String, String> params, final Optional<Duration> requestQueryTimeout) throws Exception {
     final Timer.Context ctx = registry.timer("fauna-request").time();
-    final CompletableFuture<FullHttpResponse> rv = new CompletableFuture<>();
+    final CompletableFuture<HttpResponse<String>> rv = new CompletableFuture<>();
 
-    request.headers().add("Authorization", authHeader);
-    request.headers().set("X-FaunaDB-API-Version", API_VERSION);
+    URI requestUri = URI.create(mkUrl(path));
 
-    if(jvmDriver != null) {
-      request.headers().set(X_FAUNA_DRIVER, jvmDriver.toString());
+    // Encode all query parameters
+    for (Map.Entry<String,String> entry : params.entrySet()) {
+      requestUri = appendUri(requestUri, entry.getKey(), entry.getValue());
     }
 
-    // TODO: replace as follows when upgrade to Java 9 or higher:
-    // requestQueryTimeout.or(queryTimeout)
-    //   .ifPresent(timeout -> request.headers().set(X_QUERY_TIMEOUT, timeout.toMillis()));
-    queryTimeout
-      .ifPresent(timeout -> request.headers().set(X_QUERY_TIMEOUT, timeout.toMillis()));
+
+    HttpRequest.BodyPublisher bodyPublisher = HttpRequest.BodyPublishers.noBody();
+    if (body.isPresent()) {
+      byte[] jsonBody = json.writeValueAsBytes(body.get());
+      bodyPublisher = HttpRequest.BodyPublishers.ofByteArray(jsonBody);
+    };
 
     // If a query timeout has been given for the current request,
     // override the one from the Connection if any
-    requestQueryTimeout
-      .ifPresent(timeout -> request.headers().set(X_QUERY_TIMEOUT, timeout.toMillis()));
-
-    long time = getLastTxnTime();
-    if (time > 0) {
-      request.headers().set("X-Last-Seen-Txn", Long.toString(time));
-    }
-
-    request.retain();
+    Optional<Duration> queryTimeout = requestQueryTimeout.or(() -> connectionQueryTimeout);
+
+    Optional<Long> lastTxnTime = (getLastTxnTime() > 0) ? Optional.of(getLastTxnTime()) : Optional.empty();
+
+    HttpRequest.Builder requestBuilder =
+      HttpRequest.newBuilder()
+        .uri(requestUri)
+        .version(HttpClient.Version.HTTP_1_1)
+        .timeout(Duration.ofMillis(DEFAULT_REQUEST_TIMEOUT_MS))
```

Comment:
this timeout should be set based on the passed query timeout, as we shouldn't allow the X_QUERY_TIMEOUT header to be greater than this value.

Reply:
If we set the HTTP Request Timeout with the Query Timeout value, the Query Timeout exception from the server will never reach the client, since the HTTP Request Timeout exception from the client is going to be thrown always first, right? If so, is that the behaviour we want to provide?

I understand that with the way it is today, users might run in the scenario you described, which is also unexpected. That being said, similar to the comment above, we didn't modify the behaviour from the previous client here, so it might make sense to improve this in a separate ticket and not block the PR because of it?

Reply:
"based on" was a bit too ambiguous: Agree it shouldn't be equal to the query timeout, but it could e.g. be automatically set to the query timeout + 5 seconds or some other safe margin. 

Whatever else is true, we should not allow the provided query timeout to exceed this value, as the resulting behavior is surprising in a bad way.